File: backend/app/services/tree.py
"""
Tree utilities
"""
from typing import TypeVar, Generic, List, Optional

T = TypeVar('T')


class TreeNode(Generic[T]):
    """Tree node"""
    def __init__(self, value: T):
        self.value = value
        self.children: List[TreeNode[T]] = []


class Tree(Generic[T]):
    """Tree implementation"""
    def __init__(self, root_value: Optional[T] = None):
        self._root: Optional[TreeNode[T]] = None
        if root_value is not None:
            self._root = TreeNode(root_value)

    def get_root(self) -> Optional[T]:
        """Get root value"""
        return self._root.value if self._root else None

    def set_root(self, value: T) -> None:
        """Set root value"""
        self._root = TreeNode(value)

    def add_child(self, parent_value: T, child_value: T) -> bool:
        """Add child to parent"""
        parent = self._find_node(parent_value)
        if not parent:
            return False
        parent.children.append(TreeNode(child_value))
        return True

    def remove(self, value: T) -> bool:
        """Remove a node"""
        if not self._root:
            return False
        if self._root.value == value:
            self._root = None
            return True
        return self._remove_node(self._root, value)

    def _remove_node(self, node: TreeNode[T], value: T) -> bool:
        """Remove node helper"""
        for i, child in enumerate(node.children):
            if child.value == value:
                node.children.pop(i)
                return True
            if self._remove_node(child, value):
                return True
        return False

    def find(self, value: T) -> bool:
        """Find a node"""
        return self._find_node(value) is not None

    def _find_node(self, value: T) -> Optional[TreeNode[T]]:
        """Find node helper"""
        if not self._root:
            return None
        return self._find_node_helper(self._root, value)

    def _find_node_helper(self, node: TreeNode[T], value: T) -> Optional[TreeNode[T]]:
        """Find node helper"""
        if node.value == value:
            return node
        for child in node.children:
            found = self._find_node_helper(child, value)
            if found:
                return found
        return None

    def get_children(self, value: T) -> List[T]:
        """Get children of a node"""
        node = self._find_node(value)
        return [child.value for child in node.children] if node else []

    def get_parent(self, value: T) -> Optional[T]:
        """Get parent of a node"""
        if not self._root or self._root.value == value:
            return None
        return self._get_parent_helper(self._root, value)

    def _get_parent_helper(self, node: TreeNode[T], value: T) -> Optional[T]:
        """Get parent helper"""
        for child in node.children:
            if child.value == value:
                return node.value
            parent = self._get_parent_helper(child, value)
            if parent is not None:
                return parent
        return None

    def get_height(self) -> int:
        """Get tree height"""
        if not self._root:
            return 0
        return self._get_height_helper(self._root)

    def _get_height_helper(self, node: TreeNode[T]) -> int:
        """Get height helper"""
        if not node.children:
            return 1
        return 1 + max(self._get_height_helper(child) for child in node.children)

    def traverse_bfs(self) -> List[T]:
        """Breadth-first traversal"""
        if not self._root:
            return []
        result: List[T] = []
        queue = [self._root]
        while queue:
            node = queue.pop(0)
            result.append(node.value)
            queue.extend(node.children)
        return result

    def traverse_dfs(self) -> List[T]:
        """Depth-first traversal"""
        if not self._root:
            return []
        result: List[T] = []
        self._traverse_dfs_helper(self._root, result)
        return result

    def _traverse_dfs_helper(self, node: TreeNode[T], result: List[T]) -> None:
        """DFS helper"""
        result.append(node.value)
        for child in node.children:
            self._traverse_dfs_helper(child, result)

    @property
    def size(self) -> int:
        """Get tree size"""
        return len(self.traverse_bfs())

    @property
    def is_empty(self) -> bool:
        """Check if tree is empty"""
        return self._root is None

File: src/lib/tree.ts
/**
 * Tree utilities
 */

class TreeNode<T> {
  value: T
  children: TreeNode<T>[]
  constructor(value: T, children: TreeNode<T>[] = []) {
    this.value = value
    this.children = children
  }
}

export class Tree<T> {
  private root: TreeNode<T> | null = null

  constructor(rootValue?: T) {
    if (rootValue !== undefined) {
      this.root = new TreeNode(rootValue)
    }
  }

  getRoot(): T | null {
    return this.root?.value ?? null
  }

  setRoot(value: T): void {
    this.root = new TreeNode(value)
  }

  addChild(parentValue: T, childValue: T): boolean {
    const parent = this.findNode(parentValue)
    if (!parent) return false
    parent.children.push(new TreeNode(childValue))
    return true
  }

  remove(value: T): boolean {
    if (!this.root) return false
    if (this.root.value === value) {
      this.root = null
      return true
    }
    return this.removeNode(this.root, value)
  }

  private removeNode(node: TreeNode<T>, value: T): boolean {
    for (let i = 0; i < node.children.length; i++) {
      if (node.children[i].value === value) {
        node.children.splice(i, 1)
        return true
      }
      if (this.removeNode(node.children[i], value)) {
        return true
      }
    }
    return false
  }

  find(value: T): boolean {
    return this.findNode(value) !== null
  }

  private findNode(value: T): TreeNode<T> | null {
    if (!this.root) return null
    return this.findNodeHelper(this.root, value)
  }

  private findNodeHelper(node: TreeNode<T>, value: T): TreeNode<T> | null {
    if (node.value === value) return node
    for (const child of node.children) {
      const found = this.findNodeHelper(child, value)
      if (found) return found
    }
    return null
  }

  getChildren(value: T): T[] {
    const node = this.findNode(value)
    return node?.children.map(c => c.value) ?? []
  }

  getParent(value: T): T | null {
    if (!this.root || this.root.value === value) return null
    return this.getParentHelper(this.root, value)
  }

  private getParentHelper(node: TreeNode<T>, value: T): T | null {
    for (const child of node.children) {
      if (child.value === value) return node.value
      const parent = this.getParentHelper(child, value)
      if (parent !== null) return parent
    }
    return null
  }

  getHeight(): number {
    if (!this.root) return 0
    return this.getHeightHelper(this.root)
  }

  private getHeightHelper(node: TreeNode<T>): number {
    if (node.children.length === 0) return 1
    return 1 + Math.max(...node.children.map(c => this.getHeightHelper(c)))
  }

  traverseBFS(): T[] {
    if (!this.root) return []
    const result: T[] = []
    const queue: TreeNode<T>[] = [this.root]

    while (queue.length > 0) {
      const node = queue.shift()!
      result.push(node.value)
      queue.push(...node.children)
    }

    return result
  }

  traverseDFS(): T[] {
    if (!this.root) return []
    const result: T[] = []
    this.traverseDFSHelper(this.root, result)
    return result
  }

  private traverseDFSHelper(node: TreeNode<T>, result: T[]): void {
    result.push(node.value)
    for (const child of node.children) {
      this.traverseDFSHelper(child, result)
    }
  }

  get size(): number {
    return this.traverseBFS().length
  }

  get isEmpty(): boolean {
    return this.root === null
  }
}
